src/tdsops.f90

```diff
@@ -0,0 +1,682 @@
+module m_tdsops
+   use m_common, only: dp, pi
+
+   implicit none
+
+   type :: tdsops_t
+      !! Tridiagonal Solver Operators class.
+      !! Operator arrays are preprocessed in this class based on the arguments
+      !! provided. dist_fw and dist_bw are used in the first phase of the
+      !! distributed tridiagonal solver algorithm. dist_sa and dist_sc are used
+      !! in the final substutition phase. See the kernels_dist.f90 files in the
+      !! relevant backend folders.
+      !! coeff arrays define the specific rules of building the RHS
+      !! corresponding to the tridiagonal system to be solved, and used only in
+      !! the first phase of the distributed algorithm when building the RHS.
+      !! If a boundary condition is defined then coeffs_t and coeffs_b differ
+      !! from coeffs array and define the RHS rule for the first and last 4
+      !! entries in the tridiagonal system (n_halo = 4).
+      !!
+      !! This class does not know about the current rank or its relative
+      !! location among other ranks. All the operator arrays here are used when
+      !! executing a distributed tridiagonal solver phase one or two.
+      real(dp), allocatable, dimension(:) :: dist_fw, dist_bw, dist_sa, dist_sc
+      real(dp), allocatable :: coeffs(:), coeffs_t(:, :), coeffs_b(:, :)
+      integer :: n, n_halo
+      real(dp) :: alpha, delta
+   contains
+      procedure :: deriv_1st, deriv_2nd, interpl_mid, stag_der_1st, preprocess
```

Comment:
Can we add a type-bound procedure `fin` to properly release the memory and restore the parameters to default values ?

Reply:
I noticed in Incompact3D finalisation is a thing and allocatable arrays are freed before the executable terminates. I don't normally do this, and don't understand why it is important to do so. Can you explain a bit?

Reply:
Fortran should deallocate an (allocatable) array when it goes out of scope. An explicit deallocate statement can often be a "good practice" but shouldn't be necessary

Reply:
I'm tempted to opt for not bothering with deallocation to be honest. Simulation is going to terminate at the end anyways.

Reply:
Avoiding explicit deallocate statements makes the code slightly simpler. It also makes the hunt of a memory leak very difficult. If you think the code will never face a memory leak, then it is ok to avoid explicit deallocate statements.